### ISSUE TITLE
Allow changing vital arguments without UI

### ISSUE THREAD

**Issue by gudvinr** [Sonarr/Sonarr#4699]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe

While working on Sonarr installation I had major blocking issues caused by inability to change configuration without using UI:
1. I set up Sonarr behind reverse proxy on dedicated path inside Docker. Sonar itself isn't exposed behind bridge network. This leads to inaccessible UI and broken installation because you can't access UI to change base URL and you can't change base URL without accessing UI
2. After using workaround (more on that later) Sonarr by some reason appeared to be VERY slow. It takes couple of minutes to load UI and switch between pages.  
  To investigate an issue I needed to change log level. For that you should open UI (which is slow), navigate to General settings (which is slow), change configuration and save (which is slow.

### Describe the solution you'd like

I'd like to get an ability to change some essential configuration parameters, which should be accessible outside of configuration UI. That includes:
* base URL
* listen port
* bind address
* log level

These parameters might be set using either command line arguments or environment variables. If provided parameter already set through config file, cli/env parameter should take precedence to avoid locking yourself out of UI due to misconfiguration.

### Describe alternatives you've considered

To get around first issue I implemented workaround by using bootstrapped `config.xml` which contains only `<UrlBase>/sonarr</UrlBase>`.

That was lucky shot in assumption that Sonarr fills missing values by itself if none provided.

It is undocumented workaround and should not be used.

Also you can't change parameters after you started container because Sonarr will overwrite bootstrapped file.

There are no workarounds to second issue.

### Anything else?

I use docker image based on debian (bullseye) with official packages installed.

Version: 3.0.6.1196
Package Version: 3.0.6 by Team Sonarr 
Mono Version: 6.8.0.105

**Comment by gudvinr** [comment]
@bakerboy448 could you please explain how would you change one variable using `config.xml` file without manually editing it? E.g. from docker-compose when you don't have direct access to container file system?

**Comment by bakerboy448** [comment]
> It restricts you from automating deploys because you can't change connection parameters without manual intervention

this is false as well - there is no reason you cannot systematically modify the config file 

> Main problem is that you must change parameters of deployment manually and editing files inside container isn't goto method.

You're not getting it.  You don't edit the file inside the container.  You shutdown the container - edit the config.xml for sonarr on the host system - and then start the container backup.

it would appear based on your belief that you cannot access anything from the container that you either have a complete misunderstanding how docker works and/or do not have your database and config file persisting on your host and only will have any of your settings for the current container - losing everything eventually and/or are building or using some custom built self made docker container built by someone who does not understand docker.

if you can only access sonarr's database and config file from within the container and not from the host then you do not wish to retain any of your settings and will lose everything once you pull an updated container.

if you have no plans to update the container and update sonarr within docker that is not supported and goes against the entire concept of docker

**Comment by bakerboy448** [comment]
To your edit:

> Is using docker volumes to persist folder passed in -data argument to Sonarr considered "poor setup"?

that is exactly what you do. You then edit the config file in that mount.  If you do not believe you can systematically edit a config xml file programmically then I don't know what to tell you

**Comment by mrhotio** [comment]
> @bakerboy448 could you please explain how would you change one variable using `config.xml` file without manually editing it? E.g. from docker-compose when you don't have direct access to container file system?
> 
> Existing of `config.xml` doesn't solve an issue I described and I even explained why it doesn't solve it.

yq might be of help if you don't wanna manually edit the config.xml with a text editor.

**Comment by bakerboy448** [comment]
Additionally the entire UI is lipstick on the API - anything and everything done via the UI is via the API.  Docs are linked from the wiki and/or the best way is to watch the network calls in your browser dev console.


There is no realistic situation where the following would exist at the same time:
- one cannot access the webui or hit sonarr via the api
- One cannot access the config.xml file
- one would somehow have access to change the run arguments

**Comment by bakerboy448** [comment]
> From docker documentation it's clear that it's not necessary for volumes to be "mounted on the host" to retain data between container restarts. And, in fact, Docker states that Volumes are the preferred mechanism for persisting data generated by and used by Docker containers.

that's what I was saying regarding mounting on  the host  - albeit we were using different terms

bottom line - the config file is always accessible from the host machine if you are persisting the data in some form.

> The issue is that you can't change connection parameters without manual intervention or shitton of bash scripts that will replace variables in config.xml. And if you want container to be flexible enough it's necessity.

once again - false - the whole UI uses the API to do everything so as just commented you can:
- edit the config settings in the UI
- edit the config file directly
- Make the API calls to edit the config file

No plans to add config variables as command arguments.


Also since it sounds like you may be doing some sort of Kubernetes setup.  Don't forget that sqlite databases MUST be on a local non-network storage location. It will corrupt.  (And no- no plans to add other database support)

**Comment by gudvinr** [comment]
> yq might be of help if you don't wanna manually edit the config.xml with a text editor.

It's possible to use `sed`, `awk` and whatever you might suggest in custom entrypoint to implement workaround for that.  
I raised that issue because you _need_ workarounds for that.

And I specifically stated that this only necessary for deployment-specific parameters and not as general configuration interface for every argument.

**Comment by bakerboy448** [comment]
Duplicate of https://github.com/Sonarr/Sonarr/issues/311

**Comment by gudvinr** [comment]
> you can: edit the config settings in the UI

Manual intervention.

> you can: edit the config file directly

Manual intervention or custom entrypoint for replacing values in xml file.

> you can: Make the API calls to edit the config file

Also custom entrypoint is needed and you also must have running instance of Sonarr.  
Moreover, you must store default connection parameters and extract them from xml afterwards to use in subsequent calls. Because once you change them API access wouldn't work through default params.

**Comment by gudvinr** [comment]
@mrhotio Thanks, I know that it's possible to do so. But rather than doing exact same thing through custom scripts I thought it'll be beneficial for many people to have that on the application side. It could significantly lessen burden for maintaining such configurations.

> Docker images usually have some kind of initialization process/script, particularly to handle UID/GID and other 'first install' initialization steps.

@Taloth That is correct

> You should add this to your image, use yq (or equivalent) to initialize config.xml as appropriate and then exec into Sonarr.
> These aren't workarounds, but common practice in docker image maintainer space.

Using environment variables and command line arguments to initialize parameters before loading (or creating) configuration files is also common practice.  
Not only in docker image maintainer space but in software development space in general.  
Editing application files on the fly using text replacement tools _is_ workaround and it is error-prone.
